### PR TITLE
Update JobSubmitter for docker support

### PR DESCRIPTION
**Pull request by goughes**
#### Status
Tested

#### Description
We need to modify the location of Unpacker.py for agents running in Docker containers. Since the schedd runs on the host, this script must exist somewhere on the host and be available through bind mounts in the container.

Uses an "isDocker" configuration parameter to differentiate between Docker and the standard VM WMAgent deployment and change the Unpacker.py path. The isDocker variable is updated to True when a container starts via https://github.com/dmwm/CMSKubernetes/blob/master/docker/wmagent/run.sh.

#### Is it backward compatible (if not, which system it affects?)
Yes

#### Related PRs
NA

#### External dependencies / deployment changes
NA

### CODE REVIEW

**Comment by goughes** [comment]
I think this is good to merge, but I won't have time to test a final time until after the winter recess. Let's leave it until the new year.

**Comment by goughes** [comment]
I tested this with a normal VM deployment using the 1.2.8 release. It works fine, I think we could do some additional cleanup of this code if desired.

What I found was that the Unpacker location gets set to `/data/srv/wmagent/v1.2.8/sw/slc7_amd64_gcc630/cms/wmagent/1.2.8/lib/python2.7/site-packages/WMCore/WMRuntime/Unpacker.py`. I checked on some other production agents and it gets set the same there as well.

That means the Unpacker path on the else block here doesn't really exist. And I don't see how it could ever exist.
https://github.com/dmwm/WMCore/blob/b970c2b1be1cbaea0f0f893ad348cffc3bf3ad99/src/python/WMCore/BossAir/Plugins/SimpleCondorPlugin.py#L108-L109

So the path on L114 will always get set.
https://github.com/dmwm/WMCore/blob/b970c2b1be1cbaea0f0f893ad348cffc3bf3ad99/src/python/WMCore/BossAir/Plugins/SimpleCondorPlugin.py#L113-L114

We could probably just simplify this to have an if/else statement to switch between the Docker path `os.path.join(config.General.workDir + "/Docker/WMRuntime/Unpacker.py")` and the VM path `os.path.join(getWMBASE(), 'WMCore/WMRuntime/Unpacker.py')`

File: src/python/WMCore/BossAir/Plugins/SimpleCondorPlugin.py [hunk 1]
```diff
@@ -100,10 +100,16 @@ def __init__(self, config):
...
-        if os.path.exists(os.path.join(getWMBASE(),
-                                       'src/python/WMCore/WMRuntime/Unpacker.py')):
-            self.unpacker = os.path.join(getWMBASE(),
-                                         'src/python/WMCore/WMRuntime/Unpacker.py')
+        # if agent is running in a container, Unpacker.py must come from a directory
+        # on the host so the condor schedd can see it
+        # config.General.workDir should always be bind mounted to the container
+        if config.Agent.isDocker:
+            unpackerPath = os.path.join(config.General.workDir + "/Docker/WMRuntime/Unpacker.py")
+        else:
```
Comment (amaltaro) [inline src/python/WMCore/BossAir/Plugins/SimpleCondorPlugin.py]: Eric, would you know what `getWMBASE()` is when running from a container?

Comment (amaltaro) [reply src/python/WMCore/BossAir/Plugins/SimpleCondorPlugin.py]: Oh, and did you test both of these cases?

File: src/python/WMCore/BossAir/Plugins/SimpleCondorPlugin.py [hunk 2]
```diff
@@ -100,10 +100,16 @@ def __init__(self, config):
...
         self.packageDir = None
 
-        if os.path.exists(os.path.join(getWMBASE(),
-                                       'src/python/WMCore/WMRuntime/Unpacker.py')):
-            self.unpacker = os.path.join(getWMBASE(),
-                                         'src/python/WMCore/WMRuntime/Unpacker.py')
+        # if agent is running in a container, Unpacker.py must come from a directory
+        # on the host so the condor schedd can see it
+        # config.General.workDir should always be bind mounted to the container
+        if config.Agent.isDocker:
```
Comment (amaltaro) [inline src/python/WMCore/BossAir/Plugins/SimpleCondorPlugin.py]: One more comment here, how about changing it to:
`if getattr(config.Agent, "isDocker", False):`
?
With such change, you don't necessarily need to update those test files/configs

**Review by amaltaro — COMMENTED** [review]
Erik, I made a few comments in the code. Otherwise patch looks good to me.

File: src/python/WMCore/BossAir/Plugins/SimpleCondorPlugin.py (hunk 2 above)
Comment (goughes) [reply src/python/WMCore/BossAir/Plugins/SimpleCondorPlugin.py]: Would you prefer I leave out the updates to the unit test files then?

Comment (amaltaro) [reply src/python/WMCore/BossAir/Plugins/SimpleCondorPlugin.py]: Given that they are already in place, it doesn't harm to leave them now.

File: src/python/WMCore/BossAir/Plugins/SimpleCondorPlugin.py (hunk 1 above)
Comment (goughes) [reply src/python/WMCore/BossAir/Plugins/SimpleCondorPlugin.py]: I tested both the Docker and "standard" VM deployment paths

**Comment by goughes** [comment]
Did a quick retest with getattr change to verify submission still works. Things look good from my end.

**Comment by amaltaro** [comment]
Thanks, Erik

**Pull request merged**